File: frontend/src/lib/api/services/booking.service.ts
/**
 * Booking Service
 * Handles booking management
 */

import apiClient from '../client';
import { BookingCreate, Booking } from '../types';

export const bookingService = {
  /**
   * Create a new booking (public endpoint)
   */
  async create(data: BookingCreate, workspaceId: string): Promise<Booking> {
    const response = await apiClient.post<Booking>(`/bookings?workspace_id=${workspaceId}`, data);
    return response.data;
  },

  /**
   * Get all bookings
   */
  async getAll(params?: { start_date?: string; end_date?: string }): Promise<Booking[]> {
    const response = await apiClient.get<Booking[]>('/bookings', { params });
    return response.data;
  },

  /**
   * Get today's bookings
   */
  async getToday(): Promise<Booking[]> {
    const response = await apiClient.get<Booking[]>('/bookings/today');
    return response.data;
  },

  /**
   * Get upcoming bookings
   */
  async getUpcoming(days: number = 7): Promise<Booking[]> {
    const response = await apiClient.get<Booking[]>(`/bookings/upcoming?days=${days}`);
    return response.data;
  },

  /**
   * Update booking
   */
  async update(id: string, data: Partial<BookingCreate>): Promise<Booking> {
    const response = await apiClient.patch<Booking>(`/bookings/${id}`, data);
    return response.data;
  },

  /**
   * Update booking status
   */
  async updateStatus(id: string, status: Booking['status']): Promise<Booking> {
    const response = await apiClient.post<Booking>(`/bookings/${id}/status?status=${status}`);
    return response.data;
  },
};

File: Backend/app/services/booking_service.py
"""Booking service"""
from typing import Dict, Any, List
from datetime import datetime, timedelta
from supabase import Client

from app.services.base_service import BaseService
from app.models.enums import BookingStatus
from app.core.exceptions import ValidationException, ConflictException


class BookingService(BaseService):
    """Booking management service"""
    
    def __init__(self, supabase: Client):
        super().__init__(supabase, "bookings")
    
    async def create_booking(
        self, workspace_id: str, booking_data: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Create new booking with validation"""
        # Validate booking type exists
        booking_type = await self._get_booking_type(booking_data["booking_type_id"])
        
        # Check availability
        is_available = await self._check_availability(
            workspace_id,
            booking_data["booking_type_id"],
            booking_data["scheduled_at"],
            booking_type["duration_minutes"]
        )
        
        if not is_available:
            raise ConflictException("Selected time slot is not available")
        
        # Create booking
        booking = await self.create({
            **booking_data,
            "workspace_id": workspace_id,
            "status": BookingStatus.PENDING.value,
        })
        
        self.logger.info("booking_created", booking_id=booking["id"])
        return booking
    
    async def get_bookings_by_date_range(
        self, workspace_id: str, start_date: datetime, end_date: datetime
    ) -> List[Dict[str, Any]]:
        """Get bookings within date range"""
        response = (
            self.supabase.table(self.table_name)
            .select("*")
            .eq("workspace_id", workspace_id)
            .gte("scheduled_at", start_date.isoformat())
            .lte("scheduled_at", end_date.isoformat())
            .order("scheduled_at")
            .execute()
        )
        return response.data
    
    async def update_booking_status(
        self, booking_id: str, status: BookingStatus
    ) -> Dict[str, Any]:
        """Update booking status"""
        return await self.update(booking_id, {"status": status.value})
    
    async def get_today_bookings(self, workspace_id: str) -> List[Dict[str, Any]]:
        """Get today's bookings"""
        today = datetime.now().date()
        start = datetime.combine(today, datetime.min.time())
        end = datetime.combine(today, datetime.max.time())
        
        return await self.get_bookings_by_date_range(workspace_id, start, end)
    
    async def get_upcoming_bookings(
        self, workspace_id: str, days: int = 7
    ) -> List[Dict[str, Any]]:
        """Get upcoming bookings"""
        start = datetime.now()
        end = start + timedelta(days=days)
        
        return await self.get_bookings_by_date_range(workspace_id, start, end)
    
    async def _get_booking_type(self, booking_type_id: str) -> Dict[str, Any]:
        """Get booking type details"""
        response = (
            self.supabase.table("booking_types")
            .select("*")
            .eq("id", booking_type_id)
            .eq("is_active", True)
            .execute()
        )
        
        if not response.data:
            raise ValidationException("Invalid or inactive booking type")
        
        return response.data[0]
    
    async def _check_availability(
        self,
        workspace_id: str,
        booking_type_id: str,
        scheduled_at: datetime,
        duration_minutes: int
    ) -> bool:
        """Check if time slot is available"""
        # Check if slot falls within defined availability
        day_of_week = scheduled_at.weekday()
        time_slot = scheduled_at.time()
        
        availability_response = (
            self.supabase.table("availability_slots")
            .select("*")
            .eq("workspace_id", workspace_id)
            .eq("booking_type_id", booking_type_id)
            .eq("day_of_week", day_of_week)
            .execute()
        )
        
        if not availability_response.data:
            return False
        
        # Check for overlapping bookings
        end_time = scheduled_at + timedelta(minutes=duration_minutes)
        
        overlapping_response = (
            self.supabase.table(self.table_name)
            .select("id")
            .eq("workspace_id", workspace_id)
            .eq("booking_type_id", booking_type_id)
            .gte("scheduled_at", scheduled_at.isoformat())
            .lt("scheduled_at", end_time.isoformat())
            .neq("status", BookingStatus.CANCELLED.value)
            .execute()
        )
        
        return len(overlapping_response.data) == 0
